### ISSUE TITLE
Type narrowing on TypedDict with match expresion produces an unexpected error

### ISSUE THREAD

**Issue by rlkandela** [python/mypy#16834]
**Bug Report**

When you try to narrow down the type of a `TypedDict` by checking with a `match` expression, it produces an error when it shouldn't. But when you try an equivalent code but instead of using `match`, using `if`/`else` it works as expected.

**To Reproduce**

```python
from typing import TypedDict, Literal, Union

class EmptyTaskPayload(TypedDict):
    pass

class CheckPendingTaskSystemRebootResponse(TypedDict):
    type: Literal["SYSTEM/REBOOT"]
    payload: EmptyTaskPayload
    
class SystemVersionTaskPayload(TypedDict):
    version: str
    
class CheckPendingTaskSystemVersionResponse(TypedDict):
    type: Literal["SYSTEM/VERSION"]
    payload: SystemVersionTaskPayload

CheckPendingTaskResponse = Union[
    CheckPendingTaskSystemRebootResponse,
    CheckPendingTaskSystemVersionResponse
]

def test_match(pending_task: CheckPendingTaskResponse) -> None:
    match pending_task["type"]:
        case "SYSTEM/REBOOT":
            pass
        case "SYSTEM/VERSION":
            payload = pending_task["payload"]
            version = payload["version"]
#                                      ^ Error produced here
# main.py:28: error: TypedDict "EmptyTaskPayload" has no key "version"  [typeddict-item]

def test_if(pending_task: CheckPendingTaskResponse) -> None:
    if pending_task["type"] == "SYSTEM/REBOOT":
        pass
    elif pending_task["type"] == "SYSTEM/VERSION":
        payload = pending_task["payload"]
        version = payload["version"]
```
![Code with inlay hints](https://github.com/python/mypy/assets/28935609/a5196609-849f-4e8d-bec2-b572665e2446)
![Error given in the local editor](https://github.com/python/mypy/assets/28935609/8b9bb505-9a6d-4153-b3e5-a71c5aaec562)
[Playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=25706f0418eb84399d6cf580fb3f6028) and [Gist](https://gist.github.com/mypy-play/25706f0418eb84399d6cf580fb3f6028)

**Expected Behavior**

Expected the code to be absent of errors since it's correct.

**Actual Behavior**

Produces an error in the index of `payload` with the key `"version"` saying that the other type (`EmptyTaskPayload`) does not have that key. Which is strange as the type system givs `payload` the correct type (`SystemVersionTaskPayload`) and in the alternative version with `if`/`else` it works.

**Your Environment**

- Mypy version used: 1.8.0
- Python version used: 3.10.12

**Comment by erictraut** [comment]
I don't think mypy currently supports discriminated unions of TypedDicts for pattern matching, so this isn't really a bug. It's a reasonable feature request, however. FWIW, [pyright](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjADRQAySMxIAhgDb0CqKSYKALAAoEVQ5sAzpKgBRbPAJSA1gAU2cDmDakAFEQRlKNAJQAuEVCtQEUySLETpUAMIALYlTXEUpVGiVJZQBlOEkWCAAlYgAjMDAYaMkcFElifRIKahhzS2t4QzNGZlZOAG0AImCATWCCWQBZAHpI2QAhAHkOggqAXTyrW01tUiL5BEUVdWGdAahHOyhQ8OIIADVWSX4UQLUNLR0MwyzTC2FrKAA3Te2i8JA5hed3T29ff13liI2QLYEklJpI5GbK5c75EhFJgsdgcSo1OqNJprWSRYIASQ6ADk%2BnMhgdRkswt8bgJdtMCQ5hC8vKofH50LsAQI0lAALxQXjbMpzGlvBkBFRfVbROIJZmpYi0XkeWn0j5C4mrH5-FAStIifqiYSkYjAWDEcIAfQgbBgVDcukM73QRpgKiKfLpNsFQXVxBMUAAtAA%2BKBYgTEM4XU3mtw2eW2%2B1BSoFYh9YMXaxUKTEKBVWr1ZqtTrdCqJpNJ2zSOYXFOsjOI5ootGYnEFwvWfEjdkRl12lSVZs6XHgxvWa6-bat7ukSqD1W9gDE-dnc-nAD05CBwCAbOBSABXKhkKAeEDEEQz02oAB0CDgZgATAAOIqsVdFAwgmjp8aTIIUkYVPdSKAoMAoGUYg4HTCdth-KAyjjUg-BoL0SggLURF1fUWGNJBgCtSM0A7IJHVlfkFTdQ1AQ9b0-QDFAgzmTC2wFPDlFjEg%2BnZDlKyzFp2i6HoGwuYt7D7KBiA4OjrQY6MmIqONWLZdiEU4msMWxfNSybfYWw5cT-EYrsNJ7LV%2B3AgQR30scKmMlBeyAA) supports this feature.

**Comment by rlkandela** [comment]
Shall I close this issue and open a feature request?

**Comment by tmke8** [comment]
I think this is the same as #16286. Your issue title is maybe better though.
